sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala

```diff
@@ -434,153 +418,129 @@ class GpuDynamicPartitionDataSingleWriter(
-              // entire batch that is sorted, see a new partition, the old write status is useless
-              cachesMap.get.remove(previousPartPath)
-            } else {
-              // create a new one
-              val writer = newWriter(partPath, None, 0)
-              currentWriterStatus = new WriterStatus(writer)
-              statsTrackers.foreach(_.newPartition())
+    // We have an entire batch that is sorted, so we need to split it up by key
+    val (paths, splits) = withResource(cb) { _ =>
+      withResource(getPartitionColumnsAsTable(cb)) { partitionColumnsTbl =>
+        withResource(distinctAndSort(partitionColumnsTbl)) { distinctKeysTbl =>
+          val partitionIndexes = splitIndexes(partitionColumnsTbl, distinctKeysTbl)
+          withResource(copyToHostAsBatch(distinctKeysTbl, partDataTypes)) { cbKeys =>
+            val splits = withResource(getOutputColumnsAsTable(cb)) { outputColumnsTbl =>
+              outputColumnsTbl.contiguousSplit(partitionIndexes: _*)
+            }
+            closeOnExcept(splits) { _ =>
+              // Use the existing code to convert each row into a path. It would be nice to do this
+              // on the GPU, but the data should be small and there are things we cannot easily
+              // support on the GPU right now
+              import scala.collection.JavaConverters._
+              // paths
+              val paths = cbKeys.rowIterator().asScala.map(getPartitionPath).toArray
+              (paths, splits)
             }
-          } else {
-            // create a new one
-            val writer = newWriter(partPath, None, 0)
-            currentWriterStatus = new WriterStatus(writer)
-            statsTrackers.foreach(_.newPartition())
           }
         }
+      }
+    }
+    withResource(splits) { _ =>
+      paths.zipWithIndex.foreach { case (partPath, partIx) =>
+        val partContigTable = splits(partIx)
+        // If fall back from for `GpuDynamicPartitionDataConcurrentWriter`, we should get the
+        // saved status
+        val savedStatus = updateCurrentWriterIfNeeded(partPath, cachesMap)
 
         if (savedStatus.isDefined && savedStatus.get.tableCaches.nonEmpty) {
           // convert caches seq to tables and close caches seq
-          val subTables = convertSpillBatchesToTables(savedStatus.get.tableCaches)
+          val subTables = convertSpillBatchesToTablesAndClose(savedStatus.get.tableCaches)
           // concat the caches and this `table`
           val concat = withResource(subTables) { _ =>
             // clear the caches
             savedStatus.get.tableCaches.clear()
-            subTables += table
-            Table.concatenate(subTables: _*)
+            splits(partIx) = null
+            withResource(partContigTable) { _ =>
+              subTables += partContigTable.getTable
+              Table.concatenate(subTables: _*)
+            }
           }
           // write concat table
           if (!needSplitBatch(
```

Comment:
nit: So...?? We might concat the tables together just to split them apart again?  It might be good to rename `needSplitBatch` to be a little more descriptive as to why we need to split a batch so it is clear what is happening.

Reply:
Will look, I was focused on reducing memory waste, didn't really go into the logic much yet.

Reply:
I renamed the function and tried to do some refactoring but I am finding it really not straight forward given the current design, and I think we should punt but _absolutely_ prioritize refactoring this.